autorest/azure/devicetoken_test.go

```diff
@@ -0,0 +1,222 @@
+	sender.EmitContent(MockDeviceTokenResponse)
+	sender.EmitStatus("OK", 200)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err != nil {
+		t.Errorf("got error unexpectedly")
+	}
+}
+
+func TestDeviceTokenReturnsErrorIfSendingFails(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitErrors(1)
+	sender.SetError(fmt.Errorf("this is an error"))
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err == nil {
+		t.Errorf("failed to get err") // (expecting Error occurred while exchanging device code for token)
+	}
+}
+
+func TestDeviceTokenReturnsErrorIfServerError(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitStatus("Internal Server Error", 500)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err == nil {
+		t.Errorf("failed to get error") // (expecting Exchanging Device Code for OAuth token resulted in Status)
+	}
+}
+
+func TestDeviceTokenReturnsErrorIfCannotDeserializeDeviceToken(t *testing.T) {
+	gibberishJSON := strings.Replace(MockDeviceTokenResponse, "expires_in", ";:\"gibberish", -1)
+	sender := mocks.NewSender()
+	sender.EmitContent(gibberishJSON)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err == nil {
+		t.Errorf("failed to get err") // (expecting Error occurred while exchanging device code for token)
+	}
+}
+
+func errorDeviceTokenResponse(message string) string {
+	return `{ "error": "` + message + `" }`
+}
+
+func TestDeviceTokenReturnsErrorIfAuthorizationPending(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitContent(errorDeviceTokenResponse("authorization_pending"))
+	sender.EmitStatus("Bad Request", 400)
+	client := &autorest.Client{Sender: sender}
+
+	go func() {
+		time.Sleep(1 * time.Second)
+		sender.EmitContent(errorDeviceTokenResponse("access_denied"))
+	}()
+
+	_, _ = WaitForUserCompletion(client, TestClientID, deviceCode())
+}
+
+func TestDeviceTokenReturnsErrorIfSlowDown(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitContent(errorDeviceTokenResponse("slow_down"))
+	sender.EmitStatus("Bad Request", 400)
+	client := &autorest.Client{Sender: sender}
+
+	go func() {
+		time.Sleep(1 * time.Second)
+		sender.EmitContent(errorDeviceTokenResponse("access_denied"))
+	}()
+
+	_, _ = WaitForUserCompletion(client, TestClientID, deviceCode())
+}
+
+func TestDeviceTokenReturnsErrorIfAccessDenied(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitContent(errorDeviceTokenResponse("access_denied"))
+	sender.EmitStatus("Bad Request", 400)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := WaitForUserCompletion(client, TestClientID, deviceCode())
+	if err == nil {
+		t.Errorf("failed to get error")
+	}
+	if err != ErrDeviceAccessDenied {
+		t.Errorf("got wrong error")
+	}
```

Comment:
I believe it is enough to assert the correct error is returned. Nil will be, by definition, in that case "not a correct" error. Ditto throughout.